Native GD32 Support for Creality V4.2.2

bmourit opened the pull request:
### Description

This provides native hardware Marlin HAL support for Creality V4.2.2 Boards using the GD32F303RET6 MCU.
It uses the MFL library and MFL Arduino Core that were developed in tandem with the Marlin HAL.

### Requirements

Creality V4.2.2 board with GD32F303RET6 MCU.

Note:
Ender3 has been tested, however other machines with this board should all have the same pinout and therefor be supported. Other machines will need to be tested first and/or be marked as experimental. The same goes for custom hardware changes such as LCD, probes etc. Although I am using this now with MicroProve V2 without issue, LCD changes may require extra code to support. The default 128x64 LCD that comes with Ender3 is fully supported.

### Benefits

Brings native GD32F303RET6 MCU support, which addresses the differences between STM32F103RET and this clone MCU.

Some of the main benefits are:
 - 120MHz CPU clock
 - Faster GPIOs
 - Proper voltage and stabilization bits are configured at startup
 - ADC Hardware oversampling (16x default)
 - Removes wasted function calls for setting bits with STM32 that do not exist in GD32
 - FPU support for CPU offloading

Binary code size and RAM usage are nearly identical to the STM32 HAL build, and in some cases may even be smaller, thanks to features in the latest ARM GCC package from PlatformIO

### Configurations

Configurations will be very similar to the ones currently used, with the motherboard and serial port bing the two main things requiring change. STM32 Serial starts at 1, MFL GD32 starts at 0.

We are building with -std=gnu++23 which causes a number of deprecation warnings for Marlin. Ideally these would get their own pull request if desired, so for now we build with -Wno-deprecated-declarations to suppress the warnings.

### Related Issues

Requires [merge of #35](https://github.com/MarlinFirmware/U8glib-HAL/pull/35)

bnmguy committed:
Add initial native GD32 support for Creality Ender 3 machines with GD32F303RET6 mcu using MFL library

bnmguy committed:
Prepare for initial release

bmourit committed:
Merge branch 'MarlinFirmware:bugfix-2.1.x' into bugfix-2.1.x

bnmguy committed:
fix comment about 16x oversampling

bmourit committed:
Merge branch 'MarlinFirmware:bugfix-2.1.x' into bugfix-2.1.x

bmourit committed:
Merge branch 'MarlinFirmware:bugfix-2.1.x' into bugfix-2.1.x

bnmguy committed:
Prep work for upstream. Formatting, style and small code changes in prepatation work before upstream pull request.

bnmguy committed:
Use single precision for all float types to offload the work to the FPU.

bmourit committed:
Merge branch 'MarlinFirmware:bugfix-2.1.x' into bugfix-2.1.x

bmourit committed:
Merge branch 'MarlinFirmware:bugfix-2.1.x' into bugfix-2.1.x

bmourit committed:
Merge branch 'MarlinFirmware:bugfix-2.1.x' into bugfix-2.1.x

bnmguy committed:
use latest platform changes

bnmguy committed:
Fix timer_to_index function and other timer bugs

bmourit committed:
Merge branch 'MarlinFirmware:bugfix-2.1.x' into bugfix-2.1.x

bmourit committed:
Merge branch 'MarlinFirmware:bugfix-2.1.x' into bugfix-2.1.x

bmourit committed:
Merge branch 'MarlinFirmware:bugfix-2.1.x' into bugfix-2.1.x

bnmguy committed:
update platform release url

bnmguy committed:
use existing offset_and_rename script. Only difference is we don't use encrypt and it sets an STM32 specific define, which we can just ignore.

bnmguy committed:
Fix wrong script being removed.

bnmguy committed:
Use Marlin u8glib. This will need updated if the version changes after our u8glib changes are merged

bmourit committed:
Merge branch 'MarlinFirmware:bugfix-2.1.x' into upstream_pr

bnmguy committed:
Remove pins debugging, since we dont support this feature yet

ellensp commented:
Why are you not using standard pin name designation P{A-E}{0-15}  as in every stm32 and the datasheet https://file.elecfans.com/web2/M00/2D/A4/pYYBAGHmFm-ABL9gABMJuXIE-eY432.pdf ?

bnmguy committed:
Fix pins format test failure

bnmguy committed:
Add ci build test

bmourit commented:
> Why are you not using standard pin name designation P{A-E}{0-15} as in every stm32 and the datasheet https://file.elecfans.com/web2/M00/2D/A4/pYYBAGHmFm-ABL9gABMJuXIE-eY432.pdf ?

Yes, I am far to intimate with that Datasheet, as I wrote the low level library for this as well. The names are not meaningful in the context of the peripheral library due to the way the ports and pins are set in Hardware. It also simplified things (for me) and removed a lot of mapping back and forth to use pin numbers directly. That said, it is very much so on the todo list for the Arduino Core.

If this is something you would rather wait to have before merging support in Marlin, that is perfectly fine and understandable for readability and user friendliness alone.
It should be noted though that it appears that not all newer Arduino Cores follow this pin name scheme in their cores. Some define only analog pins Ax, others both, others use custom naming.

Since the main purpose of this core (so far) is use in Marlin, your input is greatly valued. I'm not sure when I will get to it, but I will move it to the top of the priority list if this will block merging.

Thanks so much!!

RF47 commented:
will it work on creality v4.2.7?

ellensp commented:
This is why I wanted to use standard pins names.  So you can use current pins files 
the 422 and 427 are identical except the the step and direction IO pins are swapped over for each axis 
So one way or another, yes this will work for the 437 also.  (with appropriate pin file changes)

RF47 commented:
> This is why I wanted to use standard pins names. So you can use current pins files the 422 and 427 are identical except the the step and direction IO pins are swapped over for each axis So one way or another, yes this will work for the 427 also. (with appropriate pin file changes)

Thanks

Nuck-TH commented:
I second request for changing pins names back to standart PXn. This arduino like numbering is package dependent and have no place here. It is already a massive headache with atmega mcus, so please no.

I also don't get what conversion overhead you mean. It is all usually done with C preprocessor via macro defines, after which there is no runtime overhead at all.

bmourit commented:
> This is why I wanted to use standard pins names. So you can use current pins files the 422 and 427 are identical except the the step and direction IO pins are swapped over for each axis So one way or another, yes this will work for the 427 also. (with appropriate pin file changes)

Makes perfect sense. I'll add this!

bmourit commented:
> I second request for changing pins names back to standart PXn. This arduino like numbering is package dependent and have no place here. It is already a massive headache with atmega mcus, so please no.
> 
> I also don't get what conversion overhead you mean. It is all usually done with C preprocessor via macro defines, after which there is no runtime overhead at all.

There is ni Arduino-like numbering. They just use expect an actual pin number the fits in pin_size_t. Its also not changing it "back". This is a completely different Arduino core that is very, very young. So different use cases are bound to appear. That said, I def have favoritism to Marlin for accommodation.

bmourit commented:
> I second request for changing pins names back to standart PXn. This arduino like numbering is package dependent and have no place here. It is already a massive headache with atmega mcus, so please no.
> 
> I also don't get what conversion overhead you mean. It is all usually done with C preprocessor via macro defines, after which there is no runtime overhead at all.

Also, see STMduino. They use a number of conversions and they have their reasons, but digitalToPinName and the like are not completely immune to overhead depending on usage. Then there are Macros... yuck. Kidding - they have their place for sure, but debugging with many different macros can quickly become hell. Anyway, looking at the stm32duino cores it seems they actually just use alias macros for these as well, from their own naming.

bnmguy committed:
Use common pin nameing scheme

bmourit commented:
> This is why I wanted to use standard pins names. So you can use current pins files the 422 and 427 are identical except the the step and direction IO pins are swapped over for each axis So one way or another, yes this will work for the 427 also. (with appropriate pin file changes)

I have made the changes, however they are only minimally useful as it appears sharing the pins files wont be possible without breaking env_validate.h since these are different targets. I basically followed what HC32 had done before me.

ellensp commented:
just add something like in /home/ellensp/code/Marlin2/Marlin-stock/Marlin/src/pins/lpc1769/pins_SMOOTHIEBOARD.h

#define ALLOW_LPC1768_OR_9
#include "env_validate.h"

where env_validate.h has 
#if ENABLED(ALLOW_LPC1768_OR_9) && NOT_TARGET(MCU_LPC1768, MCU_LPC1769)
  #error "Oops! Make sure you have the LPC1768 or LPC1769 environment selected in your IDE."
#elif NOT_TARGET(MCU_LPC1769)
  #error "Oops! Make sure you have the LPC1769 environment selected in your IDE."
#endif

bmourit commented:
Thanks!, will do. 🙂

bnmguy committed:
Use existing board pin file

bnmguy committed:
Fix yaml syntax error

bnmguy committed:
fixed missing endif

bnmguy committed:
Use env name consistant with other boards and test case

bnmguy committed:
fix env_validate logic

classicrocker883 commented:
fine job youre doing, if i may chime in... the Voxelab Aquila uses a similar chip - GD32F103RC. and board - same pinout as 422.

a brief look at the datasheet and its almost identical to the F303. i have been able to flash the maple variant on my 427. have you tried using the env `STM32F103RC_creality_maple`? 

i know one is M3 and the other M4 cortex, im just curious if you tried using that.

bnmguy committed:
Fix emergency parser. Add default configuration for testing

bmourit commented:
> fine job youre doing, if i may chime in... the Voxelab Aquila uses a similar chip - GD32F103RC. and board - same pinout as 422.
> 
> a brief look at the datasheet and its almost identical to the F303. i have been able to flash the maple variant on my 427. have you tried using the env `STM32F103RC_creality_maple`?
> 
> i know one is M3 and the other M4 cortex, im just curious if you tried using that.

Yes, the GD32F103RC is very similar. The main difference is the smaller flash size and no FPU. I haven't tried the maple version of that on this board, no. I do want to add these variants to the Arduino Core though! The GD32F303RET6 sets some GD32 specific bits that I would need to make sure are the same or compatible with that chip though. And the FPU would need to be configured as off, since M3 doesn't have one, but this check is done in the platformio script already anyway.

This is something I really want to add though and it shouldn't be too complex with how similar they are. I'd love to see Native support for all the GD32 boards. It probably wont be in this initial PR, but a followup one.

bmourit commented:
I've added support for this board (Creality v4.2.2) and Aquila boards with this same pinout that use GD32F103RET6 and GD32F103RCT6 variants. Added to the Arduino Core dev branch and platform dev branch. Is it okay to include them along with this pull request? Or would a separate one be preferred? I've also fixed the failing tests, btw.

bmourit committed:
Merge branch 'MarlinFirmware:bugfix-2.1.x' into upstream_pr

bnmguy committed:
Add fans software pwn the test

bnmguy committed:
Fix emergency parser callback and state

bnmguy committed:
Rename gd32F3.ini to gd32. Use platform_mfl V1.0.2. Add -Wno-volatile flag to supress warnings with GCC 14

bmourit committed:
Merge branch 'MarlinFirmware:bugfix-2.1.x' into upstream_pr

thinkyhead committed:
[cron] Bump distribution date (2025-03-21)

thinkyhead committed:
🩹 Cleanup (for STM32 I2C LCD)

In advance of #26433

ellensp committed:
🩹 Fix TOUCH_UI_FTDI_EVE build (#27667)

Followup to #27275

ellensp committed:
🩹 Fix Simulator KILL pin overlap (#27668)

ellensp committed:
📌 Clean up ESP32 build (#27672)

bmourit committed:
Merge branch 'bugfix-2.1.x' into upstream_pr

bnmguy committed:
fix pin formatting from upstream

bmourit committed:
Merge branch 'bugfix-2.1.x' into upstream_pr

bnmguy committed:
Add pinsDebug.h and enable the test

bnmguy committed:
Remove MinSerial's dependancy MinSerialHelper.h. Eliminate templates and related overhead and use direct registers

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/27744

thinkyhead committed:
merge followup, etc.

thinkyhead commented on Marlin/src/pins/pins.h:
```diff
@@ -964,6 +964,13 @@
+
+#elif MB(MFL_CREALITY_V422)
+  #include "gd32f3/pins_MFL_CREALITY_V422.h"        // GD32F303RE                           env:GD32F303RE_creality_mfl
+
```
If this new environment can also be used to build Marlin for existing `CREALITY_V422` boards that happen to ship with a GD32 chip then we can also add `env:GD32F303RE_creality_mfl` to the list of environments on the `#include` line for `CREALITY_V422`. But if that would be confusing, nevermind!

thisiskeithb replied:
> we can also add `env:GD32F303RE_creality_mfl` to the list of environments on the #include line for `CREALITY_V422`

This would be preferred since it matches existing ways of doing things with `*_maple` envs.

thinkyhead replied:
If there's no specially marked "MFL" 4.2.2 board but they ship with a variety of processors then we may want to remove this added board to clarify. We can also apply the env to Aquila and others as time goes on. The environment can override the board name to add "MFL."

There is something to be said for using a set of different board names when a board has such variants. With that in mind, let's keep this one, but clarify the name….

bmourit replied:
Sounds great to me. That did also cross my mind yesterday and I meant to ask. Any way you would like to approach this is perfectly fine. Also any name changes preferred - not a problem at all. Even if it is in the Arduino core itself, we'd like to accommodate Marlin.

thinkyhead commented:
Overall, a very nice addition! In the long run it will be nice to get rid of our somewhat hacked-in GD32 support and replace it with a more specific implementation, so the integration of the MFL Library is really great, thanks!

> We are building with -std=gnu++23 which causes a number of deprecation warnings for Marlin. Ideally these would get their own pull request if desired, so for now we build with -Wno-deprecated-declarations to suppress the warnings.

There will be a big sprint following 2.1.3 to blast through some our dusty refactoring issues with an eye towards improving upgrading, extensibility, first-time setup, LCD adapters, user experience, etc. Whatever we can do to bring the code up to speed and modern C++ standards will be helpful to the overall modernization efforts, so please submit whatever will move us in that direction.

If there are no other concerns about this PR then I will go ahead and merge it imminently.

bmourit commented:
It does seem as though at least Creality has embraced GD32, so maybe the number of boards will increase at some point. Right now I believe it is just the 4.2.2, 4.2.7, and Aquila (4.2.2 pinout) and the newest cr4nxxxxx based boards. Would it be helpful to provide a base configuration.h for these boards? You would it be preferred to have the user modify the Ender3-v2 one that is currently in the configuration repo?

thinkyhead committed:
name adjust

thinkyhead committed:
Make noise about 4.2.2 boards

thinkyhead committed:
louder

thinkyhead commented:
One thing we can do is to make noise using a `#warning` in the build because the situation is fluctuating. I think it would be a good idea to make an Ender-3 config specifically for this variant of the board, and any other machines that Creality actually ships this way. But we don't need to do much more than that in the configs repo. The loud warnings should make some difference.

classicrocker883 commented:
### I found something which needs attention
what about `BOARD_CREALITY_CR4NS // Creality CR4NS200320C13 (GD32F303RET6) as found in the Ender-3 V3 SE`

however, the chip defined in firmare uses STM32F103RET6:

```y
 #include "stm32f1/pins_CREALITY_CR4NS.h"  // STM32F1    env:STM32F103RE_creality env:STM32F103RE_creality_maple
```

so should this be switched to the new GD32F303 support? or both?

classicrocker883 commented:
After trying to build, I get these errors:
```yml
Marlin/src/HAL/GD32_MFL/Servo.cpp: In function 'void fixServoTimerInterruptPriority()':
Marlin/src/HAL/GD32_MFL/Servo.cpp:44:20: error: 'getTimerUpIRQ' was not declared in this scope
   44 |   NVIC_SetPriority(getTimerUpIRQ(TIMER_SERVO), servo_interrupt_priority);
      |                    ^~~~~~~~~~~~~
Marlin/src/HAL/GD32_MFL/Servo.cpp: In member function 'int8_t libServo::attach(int)':
Marlin/src/HAL/GD32_MFL/Servo.cpp:58:21: error: 'MAX_SERVOS' was not declared in this scope; did you mean 'HAS_SERVOS'?
   58 |   if (servoCount >= MAX_SERVOS) return -1;
      |                     ^~~~~~~~~~
      |                     HAS_SERVOS
Marlin/src/HAL/GD32_MFL/Servo.cpp: In member function 'int8_t libServo::attach(int, int, int)':
Marlin/src/HAL/GD32_MFL/Servo.cpp:67:21: error: 'MAX_SERVOS' was not declared in this scope; did you mean 'HAS_SERVOS'?
   67 |   if (servoCount >= MAX_SERVOS) return -1;
      |                     ^~~~~~~~~~
      |                     HAS_SERVOS

```

bmourit commented:
This board is not currently supported in the GD32 HAL or core. I can look into adding it though

> After trying to build, I get these errors:
> 
> ```yaml
> Marlin/src/HAL/GD32_MFL/Servo.cpp: In function 'void fixServoTimerInterruptPriority()':
> Marlin/src/HAL/GD32_MFL/Servo.cpp:44:20: error: 'getTimerUpIRQ' was not declared in this scope
>    44 |   NVIC_SetPriority(getTimerUpIRQ(TIMER_SERVO), servo_interrupt_priority);
>       |                    ^~~~~~~~~~~~~
> Marlin/src/HAL/GD32_MFL/Servo.cpp: In member function 'int8_t libServo::attach(int)':
> Marlin/src/HAL/GD32_MFL/Servo.cpp:58:21: error: 'MAX_SERVOS' was not declared in this scope; did you mean 'HAS_SERVOS'?
>    58 |   if (servoCount >= MAX_SERVOS) return -1;
>       |                     ^~~~~~~~~~
>       |                     HAS_SERVOS
> Marlin/src/HAL/GD32_MFL/Servo.cpp: In member function 'int8_t libServo::attach(int, int, int)':
> Marlin/src/HAL/GD32_MFL/Servo.cpp:67:21: error: 'MAX_SERVOS' was not declared in this scope; did you mean 'HAS_SERVOS'?
>    67 |   if (servoCount >= MAX_SERVOS) return -1;
>       |                     ^~~~~~~~~~
>       |                     HAS_SERVOS
> Marlin/src/HAL/GD32_MFL/Servo.cpp:59:16: error: 'servo_pin' was not declared in this scope; did you mean 'servoPin'?
>    59 |   if (pin > 0) servo_pin = pin;
>       |                ^~~~~~~~~
>       |                servoPin
> ```
> 
> I may be able to solve one or two of these errors:
> 
> * `servoPin` in **HAL/GD32_MFL/Servo.h** should be `servo_pin`
> * Add `#define MAX_SERVOS 3` (or 12?)
> 
> which leaves `getTimerUpIRQ` was not declared, and doesn't seem to be defined anywhere. (**.platformio/packages/framework-arduino-mfl/cores/arduino/GeneralTimer.h**)

This board is not currently supported in the GD32 HAL or core. I can look into adding it though

bmourit commented:
> After trying to build, I get these errors:
> 
> ```yaml
> Marlin/src/HAL/GD32_MFL/Servo.cpp: In function 'void fixServoTimerInterruptPriority()':
> Marlin/src/HAL/GD32_MFL/Servo.cpp:44:20: error: 'getTimerUpIRQ' was not declared in this scope
>    44 |   NVIC_SetPriority(getTimerUpIRQ(TIMER_SERVO), servo_interrupt_priority);
>       |                    ^~~~~~~~~~~~~
> Marlin/src/HAL/GD32_MFL/Servo.cpp: In member function 'int8_t libServo::attach(int)':
> Marlin/src/HAL/GD32_MFL/Servo.cpp:58:21: error: 'MAX_SERVOS' was not declared in this scope; did you mean 'HAS_SERVOS'?
>    58 |   if (servoCount >= MAX_SERVOS) return -1;
>       |                     ^~~~~~~~~~
>       |                     HAS_SERVOS
> Marlin/src/HAL/GD32_MFL/Servo.cpp: In member function 'int8_t libServo::attach(int, int, int)':
> Marlin/src/HAL/GD32_MFL/Servo.cpp:67:21: error: 'MAX_SERVOS' was not declared in this scope; did you mean 'HAS_SERVOS'?
>    67 |   if (servoCount >= MAX_SERVOS) return -1;
>       |                     ^~~~~~~~~~
>       |                     HAS_SERVOS
> Marlin/src/HAL/GD32_MFL/Servo.cpp:59:16: error: 'servo_pin' was not declared in this scope; did you mean 'servoPin'?
>    59 |   if (pin > 0) servo_pin = pin;
>       |                ^~~~~~~~~
>       |                servoPin
> ```
> 
> I may be able to solve one or two of these errors:
> 
> * `servoPin` in **HAL/GD32_MFL/Servo.h** should be `servo_pin`
> * Add `#define MAX_SERVOS 3` (or 12?)
> 
> which leaves `getTimerUpIRQ` was not declared, and doesn't seem to be defined anywhere. (**.platformio/packages/framework-arduino-mfl/cores/arduino/GeneralTimer.h**)

These do not need changed. See previous response. The include path just gets corrupted, but the updated U8glib HAL fixes these errors.

ellensp commented:
BOARD_CREALITY_V422_GD32_MFL with a GD32F303 
If you enable the bltouch (which is a very common add on)  Ie  #define BLTOUCH  
Triggers the above error, as a bltouch is a  servo + probe.

This is very much meant to be a supported board.

bmourit commented:
> BOARD_CREALITY_V422_GD32_MFL with a GD32F303 If you enable the bltouch (which is a very common add on) Ie #define BLTOUCH Triggers the above error, as a bltouch is a servo + probe. No LCD of any type is enabled.
> 
> This is very much meant to be a supported board.

### 
See: https://github.com/MarlinFirmware/Marlin/pull/27791

shumnyj commented:
@bmourit Hello sir. I am currently trying to port marlin for Easythreed K10 which is build on GD32F303RCT6. But I have no experience in this area, so I got no clue what to do beyond pins/config. I saw you did great job with GD32 here and for PlatformIO, maybe you could give me some advice how should I approach it? Any help is appreciated.
